### DISCUSSION TITLE
Unable to send HTTP Request to localhost

### DISCUSSION THREAD

**Discussion by nickthiru** [Moddable-OpenSource/moddable#972] [Q&A] [answered]
Hello,

I have a local instance of Node-RED running on `localhost:1880`. I created a "http in" node in Node-RED, which I have tested with Postman. However, nothing is happening when I try to send a http request, via Moddable, when I run `mcrun -d -m -p esp32/nodemcu`. I had already successfully installed the host using `mcconfig` and there is no issue connecting to Wi-Fi (including when I run `mcrun`). Below is the code:

```
import { Request } from "http";

let request = new Request({
	host: "http://localhost:1880",
	path: "/mytopic?temp=33&humidity=44",
	response: String
});

request.callback = function (msg, value) {
	if (Request.responseComplete == msg)
		trace(value, "\n");
}
```
For the `host` property, I have tried using the following as well:

1. `localhost:1880`
2. `127.0.0.1:1880`
3. `http://127.0.0.1:1880`

Note: I have also tried adding a trailing forward slash to the above, just in case.

Thank you.

**Comment by phoddie** [comment, marked as answer]
You are close! 

- the `host` property is only the host name (e.g. `localhost`) without the protocol ("http") or port
- The port number is passed on the 'port' property 

You have find the examples in the [documentation for the HTTP request constructor](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/network.md#constructordictionary-2)  helpful, such as:

```js
let request = new Request({host: "www.example.com", path: "/info.dat", port: 8080, response: ArrayBuffer});
```

**Reply by nickthiru** [reply]
@phoddie Hey! Good to hear from you.

Thanks for pointing out the "port" property; I don't know how I could have missed that! I have yet to cover the documentation in depth, so my apologies for the "noob" mistake.

I tried adding the "port" property but the node wasn't getting "hit". (I realize that this issue has an overlap with Node-RED, and isn't strictly about Moddable, but I assumed you may be able to help since you created the "node-red-mcu" implementation (https://github.com/phoddie/node-red-mcu) :-), which I'm looking forward to studying soon!). Please find my implementation below:

Code:
![image](https://user-images.githubusercontent.com/47299208/205609815-ca24ef75-6aed-4c6f-96b8-96242dd2ca17.png)

Node-RED:
![image](https://user-images.githubusercontent.com/47299208/205610056-f9e929d8-6971-4936-8d94-34238feacc98.png)

Postman (successful request):
![image](https://user-images.githubusercontent.com/47299208/205612237-bec128e2-77e5-4b61-83e2-bffb5d71300a.png)

Extras (results from Postman request):

![image](https://user-images.githubusercontent.com/47299208/205613412-08c6f6a6-7bd8-40b2-9fa4-5e19e9b21a8d.png)

![image](https://user-images.githubusercontent.com/47299208/205613562-d29c95e9-2905-49c8-9f32-44c0ade17210.png)

Note: I tried using `localhost` as the value of the `host` property as well.

**Reply by phoddie** [reply]
I've seen macOS fail to resolve `"localhost"` in some cases. (It is strange. In modSocket.m `CFHostGetAddressing` returns NULL for the array but the `resolved` Boolean indicates resolution succeeded. Not sure why that is.)

An easy workaround is to use your local IP address (10.0.1.21) instead of localhost. If you do that, you should pass it using the `address` property instead of `host`.

**Reply by nickthiru** [reply]
Thanks for the additional information. 

I tried using the `address` property and setting it with the IP address of `localhost` but it still didn't work, unfortunately.

![image](https://user-images.githubusercontent.com/47299208/205853899-4a40d80e-4e5d-4427-b8c4-6bb54f213e4f.png)


Note: I am using Windows 11.

**Reply by Frida854** [reply]
Just a thought, have you tried removing the last comma in "String," maybe that's it.

**Reply by nickthiru** [reply]
@Frida854 Hi. Unfortunately, that didn't work either.

**Reply by Frida854** [reply]
Try putting 1880 in quotes "1880"
I just set it all up on my mac, but I have my node-red on an rpi, so a different address. It all worked at first glance.
```
import { Request } from "http";

let request = new Request({
	host: "192.168.5.85",
	port: "1880",
	path: "/mytopic?temp=33&humidity=44",
	response: String
});
```

**Reply by nickthiru** [reply]
Still no luck, I'm afraid :-( , but thanks anyway.

Note: I wanted to create an (free) instance of Node-RED on IBM Cloud, which will provide me with a static IP/host URL that ought to work (based on other examples I have seen), assuming the issue has to do with using `localhost` or its IP address. However, I haven't got round to doing this yet.

**Reply by andycarle** [reply]
I just tried a more basic test of interacting between the Moddable SDK Simulator and Node-RED via `localhost` on Windows. That seems to be working as expected, for what it's worth.

My test was:

1. I ran the Node-RED server on my Windows machine.
2. I edited the [`%MODDABLE%\examples\network\http\httpget` example ](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/network/http/httpget/main.js) to change the `Request` to: `let request = new Request({host: "localhost", path: "/", response: String, port: 1880});`
3. I ran that example in the Windows simulator. In the xsbug console, I got:

```console
<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=0"/>
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="mobile-web-app-capable" content="yes">
<!--
  Copyright OpenJS Foundation and other contributors, https://openjsf.org/

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
<title>Node-RED</title>
<link rel="icon" type="image/png" href="favicon.ico">
<link rel="mask-icon" href="red&#x2F;images&#x2F;node-red-icon-black.svg" color="#8f0000">
<link rel="stylesheet" href="vendor/jquery/css/base/jquery-ui.min.css?v=3.0.2">
<link rel="stylesheet" href="vendor/font-awesome/css/font-awesome.min.css?v=3.0.2">
<link rel="stylesheet" href="red/style.min.css?v=3.0.2">
<link rel="stylesheet" href="vendor/monaco/style.css?v=3.0.2">
</head>
<body spellcheck="false">
<div id="red-ui-editor"></div>
<script src="vendor/vendor.js?v=3.0.2"></script>
<script src="vendor&#x2F;monaco&#x2F;monaco-bootstrap.js?v=3.0.2"></script>
<script src="red&#x2F;red.min.js?v=3.0.2"></script>
<script src="red&#x2F;main.min.js?v=3.0.2"></script>


</body>
</html>
```

I did have to clear a Windows Defender Firewall pop-up when running Node-RED for the first time. And I'm sure that I would have seen a Firewall pop-up the first time I ran the Moddable Simulator in a network example. You might double-check your firewall settings and make sure the Moddable Simulator is approved. You could also verify that other network examples (that communicate with the outside world) are working correctly.

This was the one triggered by Node-RED (I selected both checkboxes):
![image (32)](https://user-images.githubusercontent.com/1427817/206006121-518d9bd1-08c7-4047-8755-2e10c05e5846.png)

I know that Windows Firewall can be touchy about loopback at times. But as far as I know, that is only a major [problem with UWP apps](https://learn.microsoft.com/en-us/windows/iot-core/develop-your-app/loopback), which we're not dealing with here.

**Reply by Frida854** [reply]
You are running the program from an esp32 right?
Then you cannot use localhost or 127.0.0.1,
because it is only on windows or mac direct,
it can be done.
When sending from esp32, it must have the address
on your windows or mac machine to work.

**Reply by nickthiru** [reply]
@andycarle Thank you for the effort and detailed feedback. I will look into it further. Regarding what you mentioned about trying out the other network examples, yes I made sure the default examples from "iot-product-dev-book -> ch3-network" were working, which only made me scratch my head more haha.

@Frida854 Yes, I am using an esp32. However, I am a little unclear about what you mean. On my end, I installed node-red on Windows 11 by following the "Quick Start" section here: https://nodered.org/docs/getting-started/windows. I run node-red from the Windows 11 Command Prompt, and the node-red server starts on `http://127.0.0.1:1880/`. Does this have any bearing on what you mentioned about "only on windows or mac direct it can be done" and "it must have the address on your windows or mac machine to work"? Thanks.

**Reply by andycarle** [reply]
@nickthiru I see. We were perhaps confused about what you are trying to accomplish.

If you are running the network application on an ESP32, you will not be able to connect to `localhost` — that refers to the device itself (in this case, the ESP32). Because your server is running on your Windows machine, you will need to use the IP address of the Windows machine on your local network.

You can find your Windows machine's IP address by running `ipconfig /all` at the Command Prompt. Look for the network adapter you use to connect to your local network and note the IP address. For instance, I have my laptop connected to the network via it's dock's Ethernet adapter. And I get this from `ipconfig /all`:

```console
Ethernet adapter Ethernet 2:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Surface Ethernet Adapter
<snipped>
   Link-local IPv6 Address . . . . . : fe80::c661:25af:7e2e:2899%14(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.68.73(Preferred)
<snipped>
```

So, the address I would use to connect my ESP32 to my Windows machine would be `192.168.68.73` and the full request would be:

```javascript
let request = new Request({
	address: "192.168.68.73",
	path: "/mytopic?temp=33&humidity=44",
	response: String,
        port: 1880
});
```

You will, of course, need to substitute the IP address that you find from `ipconfig /all`.

**Reply by nickthiru** [reply]
Right, I get it now. Sorry for the confusion and lack of specifics when raising the issue. I completely missed the point that the esp32 is a separate network host, and I have been "pointing" to _its_ `localhost` all the while.

It's working fine now. But an important thing to note, still, is what @Frida854 mentioned about specifying the port value as a String type. This is also a must.

@phoddie @andycarle @Frida854 Thank you very much for your kind and patient assistance.
